neps/nep-0508.md

```diff
@@ -0,0 +1,231 @@
+* No apps, tools or code should hardcode the number of shards to 4.
+
+### Out of scope
+
+* Dynamic resharding
+  * automatically scheduling resharding based on shard usage/capacity
+  * automatically determining the shard layout
+* Merging shards or boundary adjustments
+* Shard reshuffling
+* Shard Layout determination logic. Shard boundaries are still determined offline and hardcoded.
+
+### Required protocol changes
+
+A new protocol version will be introduced specifying the new shard layout which would be picked up by the resharding logic to split the shard.
+
+### Required state changes
+
+* For the duration of the resharding the node will need to maintain a snapshot of the flat state and related columns. As the main database and the snapshot diverge this will cause some extent of storage overhead.
+* For the duration of the epoch before the new shard layout takes effect, the node will need to maintain the state and flat state of shards in the old and new layout at the same time. The State and FlatState columns will grow up to approx 2x the size. The processing overhead should be minimal as the chunks will still be executed only on the parent shards. There will be increased load on the database while applying changes to both the parent and the children shards. 
+
+### Resharding flow
+
+* The new shard layout will be agreed on offline by the protocol team and hardcoded in the neard reference implementation.
+* In epoch T, past the protocol version upgrade date, nodes will vote to switch to the new protocol version. The new protocol version will contain the new shard layout. 
+* In epoch T, in the last block of the epoch, the EpochConfig for epoch T+2 will be set. The EpochConfig for epoch T+2 will have the new shard layout.
+* In epoch T + 1, all nodes will perform the state split. The child shards will be kept up to date with the blockchain up until the epoch end first via catchup, and later as part of block postprocessing state application.
+* In epoch T + 2, the chain will switch to the new shard layout. 
+
+## Reference Implementation
+
+The implementation heavily re-uses the implementation from [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md). Below are listed the major differences and additions. 
+
+### Code pointers to the proposed implementation 
+
+* [new shard layout](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/core/primitives/src/shard_layout.rs#L161)
+* [the main logic for splitting states](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/chain/chain/src/resharding.rs#L248)
+* [the main logic for applying chunks to split states](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/chain/chain/src/chain.rs#L5180)
+* [the main logic for garbage collecting state from parent shard](https://github.com/near/nearcore/blob/eb824087fdd0763f00e695e303d7ad6c56f96538/chain/chain/src/store.rs#L2325)
```

Comment:
Links need to be updated, at least for shard layout:  https://github.com/near/nearcore/blob/master/core/primitives/src/shard_layout.rs#L161

Logic for applying chunks was recently moved as well, but it can be moved somewhere else in near future :)

Reply:
It would make sense to use the hash of the latest commit on master instead of the master branch as that can change over time. I think @wacban has already done this in his change.